packages/connect-popup/src/index.tsx

```diff
@@ -218,70 +141,215 @@ const handleMessage = (
+
+    if (!data) return;
+
+    if (disposed) return;
+
+    log.debug('handleMessage', data);
+    handleResponseEvent(data);
+
+    // This is message from the window.opener
+    if (data.type === UI_REQUEST.IFRAME_FAILURE) {
+        fail({
+            type: 'error',
+            detail: 'iframe-failure',
+        });
 
-    if (!payload.systemInfo) {
-        payload.systemInfo = getSystemInfo(config.supportedBrowsers);
+        return;
     }
 
-    const isBrowserSupported = await view.initBrowserView(payload.systemInfo);
-    if (!isBrowserSupported) {
+    // ignore messages from origin other then MessagePort (iframe)
+    const isMessagePort =
+        event.target instanceof MessagePort ||
+        (typeof BroadcastChannel !== 'undefined' && event.target instanceof BroadcastChannel);
+    if (!isMessagePort) return;
+
+    // catch first message from iframe
+    if (data.type === POPUP.HANDSHAKE) {
+        handshake(data, event.origin);
         return;
     }
 
+    const message = parseMessage(data);
+
+    analytics.report({ type: EventType.ViewChange, payload: { nextView: message.type } });
+
+    if (
+        message?.payload &&
+        typeof message.payload === 'object' &&
+        'analytics' in message.payload &&
+        message.payload.analytics
+    ) {
+        analytics.report(message.payload.analytics);
+    }
+
+    handleUIAffectingMessage(message);
+};
+
+// handle POPUP.INIT message from window.opener
+const init = async (payload: PopupInit['payload']) => {
+    log.debug('popup init', payload);
+
+    if (!payload) return;
+
     // try to establish connection with iframe
     try {
+        if (!payload) return;
+
+        if (!payload.systemInfo) {
+            payload.systemInfo = getSystemInfo(config.supportedBrowsers);
+        }
+
+        const isBrowserSupported = await view.initBrowserView(payload.systemInfo);
+        log.debug('browser supported: ', isBrowserSupported);
+        if (!isBrowserSupported) {
+            return;
+        }
+
         // render react view
-        await renderConnectUI();
-        reactEventBus.dispatch({ type: 'waiting-for-iframe-init' });
-        await initMessageChannel(payload, handleMessage);
-        reactEventBus.dispatch({ type: 'waiting-for-iframe-handshake' });
+        renderConnectUIPromise = renderConnectUI();
+        await renderConnectUIPromise;
+        log.debug('connect-ui rendered');
+
+        addWindowEventListener('message', handleMessageInIframeMode, false);
+        await initCoreInIframe(payload);
     } catch (error) {
         postMessageToParent(createPopupMessage(POPUP.ERROR, { error: error.message }));
     }
 };
 
+const initCoreInIframe = async (payload: PopupInit['payload']) => {
+    reactEventBus.dispatch({ type: 'loading', message: 'waiting for iframe init' });
+    await initMessageChannelWithIframe(payload, handleMessageInIframeMode);
+    // done, popup is ready to handle incoming messages, waiting for handshake from iframe
+};
```

Comment:
here I am already introducing method naming that will be useful once popup starts working in dual mode (core in iframe vs core in popup)

Reply:
So here it is just preparing the naming, but it is not acctually introducing any new mode?

Reply:
mostly yes. nothing really new. I wanted to reorganize code without making big changes so that the other PR with real changes is easier to review.